src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,269 @@
+     * - throws a WebAuthnAuthenticationException if some other error occourred
+     */
+    public String finishRegistration(
+            PublicKeyCredential<AuthenticatorAttestationResponse, ClientRegistrationExtensionOutputs> pkc,
+            String realm,
+            String key) throws WebAuthnAuthenticationException {
+        try {
+            final WebAuthnCredentialCreationInfo info = activeRegistrations.get(key);
+            WebAuthnAuthenticationException noSuchRegistration = new WebAuthnAuthenticationException("_",
+                    "Can not find matching active registration request");
+            if (info == null) {
+                throw noSuchRegistration;
+            }
+            if (!info.getRealm().equals(realm)) {
+                throw noSuchRegistration;
+            }
+            final String username = info.getUsername();
+            if (!StringUtils.hasText(username)) {
+                throw new WebAuthnAuthenticationException("_",
+                        "Could not finish registration: missing username");
+            }
+            WebAuthnUserAccount existingAccount = webAuthnUserAccountService.findByProviderAndUsername(provider,
+                    username);
+            if (existingAccount != null) {
+                throw new WebAuthnAuthenticationException("_",
+                        "Account already registered");
+            }
+            final PublicKeyCredentialCreationOptions options = info.getOptions();
+            RegistrationResult result = rp
+                    .finishRegistration(FinishRegistrationOptions.builder().request(options).response(pkc).build());
+            boolean attestationIsTrusted = result.isAttestationTrusted();
+            if (attestationIsTrusted) {
+                // Create user account in the database
+                WebAuthnUserAccount account = new WebAuthnUserAccount();
+                account.setUsername(username);
+                account.setRealm(realm);
+                String subject = subjectService.generateUuid(SystemKeys.RESOURCE_USER);
+                account.setSubject(subject);
+                ByteArray userHandle = info.getOptions().getUser().getId();
+                account.setUserHandle(userHandle.getBase64());
+                account.setProvider(provider);
+                webAuthnUserAccountService.addAccount(account);
+
+                // Create credential in the database
+                WebAuthnCredential newCred = new WebAuthnCredential();
+                newCred.setCreatedOn(new Date());
+                newCred.setLastUsedOn(new Date());
+                newCred.setCredentialId(result.getKeyId().getId().getBase64());
+                newCred.setPublicKeyCose(result.getPublicKeyCose().getBase64());
+                newCred.setSignatureCount(result.getSignatureCount());
+                Optional<SortedSet<AuthenticatorTransport>> transportsOpt = result.getKeyId().getTransports();
```

Comment:
rewrite the block:
- we don't need the set as sorted
- use streams to map values
- let the field null if not set

```
          if (result.getKeyId().getTransports().isPresent()) {
                    Set<AuthenticatorTransport> transports = result.getKeyId().getTransports().get();
                    List<String> transportCodes = transports.stream().map(t -> t.getId()).collect(Collectors.toList());
                    newCred.setTransports(StringUtils.collectionToCommaDelimitedString(transportCodes));
                }
```

Reply:
Done, 744314d5e8dc67386a74144499be4f65c7f0ebad